### PR TITLE
Forum move threads between categories

### PR DESCRIPTION
**Pull request by ondratra**
Solves subtask of #766:
- Allow moderators to move threads between categories, if they have power in both source and destination.

To minimize merge conflict potential, this PR assumes #882 will be merged first.

### CODE REVIEW

File: runtime-modules/forum/src/lib.rs [hunk 1]
```diff
@@ -1257,19 +1283,53 @@ impl<T: Trait> Module<T> {
...
+        ),
+        &'static str,
+    > {
+        ensure!(category_id != new_category_id, ERROR_THREAD_MOVE_INVALID,);
+
+        let info = Self::ensure_can_moderate_thread(origin, moderator_id, category_id, thread_id);
+        if info.is_err() {
+            return Err(ERROR_MODERATOR_MODERATE_ORIGIN_CATEGORY);
+        }
+        let (account_id, thread) = info.unwrap();
```
Comment (bedeho) [inline runtime-modules/forum/src/lib.rs]: This is not statically panic-free, but it can be, ill leave it for you to find the best way.

**Review by bedeho — CHANGES_REQUESTED** [review]
Almost there, otherwise good.

File: runtime-modules/forum/src/lib.rs [hunk 2]
```diff
@@ -1316,16 +1316,31 @@ impl<T: Trait> Module<T> {
         ensure!(category_id != new_category_id, ERROR_THREAD_MOVE_INVALID,);
 
         let info = Self::ensure_can_moderate_thread(origin, moderator_id, category_id, thread_id);
-        if info.is_err() {
-            return Err(ERROR_MODERATOR_MODERATE_ORIGIN_CATEGORY);
+
+        match info {
```
Comment (bedeho) [inline runtime-modules/forum/src/lib.rs]: This is indeed safe, but its a bit hard to follow, try this

```Rust
let (account_id, thread) = Self::ensure_can_moderate_thread(..)
.map_err(ERROR_MODERATOR_MODERATE_ORIGIN_CATEGORY)?;

Self::ensure_can_moderate_category(..)
.map_err(ERROR_MODERATOR_MODERATE_DESTINATION_CATEGORY)?;

Ok(..)
```

**Review by bedeho — CHANGES_REQUESTED** [review]

File: runtime-modules/forum/src/lib.rs (hunk 2 above)
Comment (ondratra) [reply runtime-modules/forum/src/lib.rs]: Much better, thx.

**Review by bedeho — APPROVED** [review]
LGTM.
